Initial version of client

hikirsch opened the pull request.

Commit:
initial commit

choweth commented on htd/__init__.py:
```diff
@@ -0,0 +1,283 @@
+
+
+class HtdClient:
+    _ip_address: str = None
+    _port: int = None
+    _command_delay_sec: float = None
+    _retry_attempts: int = None
+    _socket_timeout_sec: float = None
```
**nitpick:** I don't think you need this as they're all defined in your constructor.

choweth commented on htd/__init__.py:
```diff
@@ -0,0 +1,283 @@
+        command_delay: int = HtdConstants.DEFAULT_COMMAND_DELAY,
+        socket_timeout: float = HtdConstants.DEFAULT_SOCKET_TIMEOUT,
+    ):
+        self._ip_address = ip_address
+        self._port = port
+        self._retry_attempts = retry_attempts
+        self._command_delay_sec = command_delay / ONE_SECOND
+        self._socket_timeout_sec = socket_timeout / ONE_SECOND
```
**nitpick(cont.):** and you can still type them here
```suggestion
        self._ip_address: str = ip_address
        self._port: int = port
        self._retry_attempts: int = retry_attempts
        self._command_delay_sec: float = command_delay / ONE_SECOND
        self._socket_timeout_sec: float = socket_timeout / ONE_SECOND
```

choweth commented on htd/utils.py:
```diff
@@ -0,0 +1,164 @@
+    return bytes(cmd)
+
+
+# convert the volume into a usable value. the device will transmit a number
+# between 196 - 255. if it's at max volume, the raw volume will come as 0.
+# this is probably because the gateway only transmits 8 bits per byte.
+# 255 is 0b11111111. since there's no volume = 0 (use mute I guess), if the
+# volume hits 0, it's because it's at max volume, so we make it 256.
+# credit for this goes to lounsbrough
+def convert_volume(raw_volume: int) -> (int, int):
```
**nitpick:** imo, in-line comments are better with this simple formatting. I would use doc strings for functions so that you can parse the info if you ever care to.
```suggestion
def convert_volume(raw_volume: int) -> (int, int):
    """
    convert the volume into a usable value. the device will transmit a number
    between 196 - 255. if it's at max volume, the raw volume will come as 0.
    this is probably because the gateway only transmits 8 bits per byte.
    255 is 0b11111111. since there's no volume = 0 (use mute I guess), if the
    volume hits 0, it's because it's at max volume, so we make it 256.
    credit for this goes to lounsbrough
    """
```

Commit:
documentation

Commit:
adding docs folder

Commit:
logic errors with handling this first 14 bytes, parse everything possible in the message and return the zone we want, parse_single_zone

Commit:
0.0.3

Commit:
starting work on lync support

hikirsch committed:
lync is working, lots of cleanup needed, need to get mca working

hikirsch committed:
working with both systems

hikirsch committed:
lync and mca working

hikirsch committed:
fixing state toggles

hikirsch committed:
0.0.6

hikirsch committed:
0.0.9 trying to fix broken pipe, attempt reconnect

hikirsch committed:
0.0.10 - pyserial support, can support direct serial connection now

hikirsch committed:
0.0.11 - reconnect if didn't disconnect on purpose

hikirsch committed:
0.0.12 - reconnect if didn't disconnect on purpose, working properly.

hikirsch committed:
0.0.13 - asyncio

hikirsch committed:
0.0.14 - fixing heartbeat time.

hikirsch committed:
0.0.15 - more async, stop blocking with while loops.

hikirsch committed:
0.0.16 - fixing async

hikirsch committed:
0.0.17 - fixing async

hikirsch committed:
0.0.18 - abort volume change if turns off

hikirsch committed:
0.0.19 - fixing mute

hikirsch committed:
0.0.20 - fixing mute for mca

hikirsch committed:
0.0.21 - use zero not null when calling callback for already zone update, do not assume zone detail is there if we didn't get all zone status response

hikirsch committed:
0.22

- intercom source support for lync
- sources greater than 13 support for lync
- clear out zone buffer if error processing data, then call refresh

hikirsch committed:
0.23 -correct name for lync6

hikirsch committed:
0.24 -correct name for lync6

hikirsch committed:
disconnect and volume parse fix

hikirsch committed:
async lock

hikirsch committed:
async issues and reconnect fixes, volume fixes apply to mca too

hikirsch committed:
test coverage, few bugs found

hikirsch committed:
max volume bug

hikirsch committed:
Merge pull request #2 from hikirsch/lync

Lync and bugs and unit tests.

hikirsch committed:
removing realtest.py

hikirsch committed:
Merge pull request #3 from hikirsch/lync

removing realtest.py
